src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -2052,83 +2053,96 @@ SubquerySqlTaskList(Job *job, PlannerRestrictionContext *plannerRestrictionConte
 	uint64 jobId = job->jobId;
 	List *sqlTaskList = NIL;
 	List *rangeTableList = NIL;
-	ListCell *rangeTableCell = NULL;
+	ListCell *restrictionCell = NULL;
 	uint32 taskIdIndex = 1; /* 0 is reserved for invalid taskId */
-	Oid relationId = 0;
 	int shardCount = 0;
 	int shardOffset = 0;
-	DistTableCacheEntry *targetCacheEntry = NULL;
 	RelationRestrictionContext *relationRestrictionContext =
 		plannerRestrictionContext->relationRestrictionContext;
+	bool *taskRequiredForShardIndex = NULL;
+	List *prunedRelationShardList = NIL;
+	ListCell *prunedRelationShardCell = NULL;
+	bool isMultiShardQuery = false;
+	bool hasDistributedTable = false;
 
 	/* error if shards are not co-partitioned */
 	ErrorIfUnsupportedShardDistribution(subquery);
 
 	/* get list of all range tables in subquery tree */
 	ExtractRangeTableRelationWalker((Node *) subquery, &rangeTableList);
 
-	/*
-	 * Find the first relation that is not a reference table. We'll use the shards
-	 * of that relation as the target shards.
-	 */
-	foreach(rangeTableCell, rangeTableList)
+	if (list_length(relationRestrictionContext->relationRestrictionList) == 0)
 	{
-		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		ereport(ERROR, (errmsg("cannot handle complex subqueries when the "
+							   "router executor is disabled")));
+	}
+
+	prunedRelationShardList = TargetShardIntervalsForQuery(subquery,
+														   relationRestrictionContext,
+														   &isMultiShardQuery);
+
+	forboth(prunedRelationShardCell, prunedRelationShardList,
+			restrictionCell, relationRestrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(restrictionCell);
+		Oid relationId = relationRestriction->relationId;
+		List *prunedShardList = (List *) lfirst(prunedRelationShardCell);
+		ListCell *shardIntervalCell = NULL;
 		DistTableCacheEntry *cacheEntry = NULL;
 
-		relationId = rangeTableEntry->relid;
 		cacheEntry = DistributedTableCacheEntry(relationId);
 		if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
 		{
 			continue;
 		}
 
-		targetCacheEntry = DistributedTableCacheEntry(relationId);
-		break;
-	}
+		/* we expect distributed tables to have the same shard count */
+		if (shardCount > 0 && shardCount != cacheEntry->shardIntervalArrayLength)
+		{
+			ereport(ERROR, (errmsg("shard counts of co-located tables do not "
+								   "match")));
+		}
 
-	/*
-	 * That means all tables are reference tables and we can pick any any of them
-	 * as an anchor table.
-	 */
-	if (targetCacheEntry == NULL)
-	{
-		RangeTblEntry *rangeTableEntry = NULL;
+		if (taskRequiredForShardIndex == NULL)
+		{
+			shardCount = cacheEntry->shardIntervalArrayLength;
+			taskRequiredForShardIndex = (bool *) palloc0(shardCount);
```

Comment:
I think if we use a `bms` here instead of `taskRequiredForShardIndex `, we could even get rid of the following `for (shardOffset = 0; shardOffset < shardCount; shardOffset++)` which might be a useful win for very large shard counts?

Reply:
👍